### PR TITLE
Scaladoc diagrams (again)

### PR DESCRIPTION
**Pull request by VladUreche**
Review by:
@kzys, @heathermiller

### CODE REVIEW

File: src/compiler/scala/tools/nsc/doc/html/page/diagram/DotDiagramGenerator.scala [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+
+        // it seems dot chokes on node names over 8000 chars, so let's limit the size of the string
+        // conservatively, we'll limit at 4000, to be sure:
+        def limitSize(str: String) = if (str.length > 4000) str.substring(0, 3996) + " ..." else str
+
+        // avoid overcrowding the diagram:
+        //   if there are too many super / sub / implicit nodes, represent
+        //   them by on node with a corresponding tooltip
+        superClasses = if (_superClasses.length > settings.docDiagramsMaxNormalClasses.value) {
+          val superClassesTooltip = Some(limitSize(_superClasses.map(_.tpe.name).mkString(", ")))
```
Comment (heathermiller) [inline src/compiler/scala/tools/nsc/doc/html/page/diagram/DotDiagramGenerator.scala:88]: Note to self: we should change this to a modal box with links instead.

File: src/compiler/scala/tools/nsc/doc/html/page/diagram/DotRunner.scala [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    val tFinish = System.currentTimeMillis
+    DiagramStats.addDotRunningTime(tFinish - tStart)
+
+    if (result == null) {
+      dotProcess.cleanup()
+      dotProcess = null
+      if (dotRestarts == settings.docDiagramsDotRestart.value) {
+        settings.printMsg("\n")
+        settings.printMsg("**********************************************************************")
+        settings.printMsg("Diagrams will be disabled for this run beucause the graphviz dot tool")
```
Comment (heathermiller) [inline src/compiler/scala/tools/nsc/doc/html/page/diagram/DotRunner.scala:49]: beucause? (Sorry for commenting on typos, but this one actually gets printed)

File: src/compiler/scala/tools/nsc/doc/model/Entity.scala [hunk 3]
```diff
@@ -446,12 +497,30 @@ trait ImplicitConversion {
   def members: List[MemberEntity]
 }
 
-/** A trait that encapsulates a constraint necessary for implicit conversion */
-trait Constraint {
-  // /** The implicit conversion during which this constraint appears */
-  // def conversion: ImplicitConversion
+/** Shadowing captures the information that the member is shadowed by some other members
```
Comment (heathermiller) [inline src/compiler/scala/tools/nsc/doc/model/Entity.scala:500]: Excellent!!

File: src/library/scala/concurrent/SyncVar.scala [hunk 4]
```diff
@@ -72,25 +75,28 @@ class SyncVar[A] {
   }
 
   // TODO: this method should be private
-  // [Heather] the reason why: it doesn't take into consideration
```
Comment (heathermiller) [inline src/library/scala/concurrent/SyncVar.scala:75]: Thanks for fixing the whitespace problems in this file :)

**Comment by heathermiller** [comment]
Thanks a lot Vlad! LGTM
We still have TODOs before 2.10, mostly regarding usability, but this looks really excellent! Nice going on better-organizing the model, the way you organized it makes a lot of sense.

**Comment by heathermiller** [comment]
Out of curiosity, could you point me to what you did to solve ticket SI-5212? (I ask because if there's an automatic js redirect, one could wind up in a mess where there are iframes embedded within iframes embeded within iframes, etc...)

**Comment by heathermiller** [comment]
Also, sure, the raw output looks like a more desirable way to test. However, could you clarify what workflow-change that you mean in the commit message (in 44ec110) when you say, 

> "There is a cool project by Stefan Zeiger to export
> the scaladoc model in JSON, but with the language and scaladoc being so
> quick to evolve, it'll be a pain to properly maintain. In the long-run,
> the plan is to sample a couple of raw files on each build and email me
> the diff. If I spot anything that may be wrong I can fix it, revert the
> commit or at least file a bug."

File: src/compiler/scala/tools/nsc/doc/html/page/diagram/DotRunner.scala (hunk 2 above)
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/doc/html/page/diagram/DotRunner.scala:49]: Fixed, will be there in the next pull request batch.

**Comment by VladUreche** [comment]
Regarding ticket SI-5212: Damien fixed it, the change is here: https://github.com/damienobrist/scala/commit/5efb003605bbe86d91a424e96b88cbe901655f4e

**Comment by VladUreche** [comment]
Raw output has 2 main drawbacks:
- it depends on the html page that gets generated, so any layout change significantly changes the raw output (remember the HTML tests? They test the raw output, and any layout change breaks them)
- it won't capture all the aspects of the model, such as the signatures that entities receive or the diagram contents. (as they're embedded in html elements) 

On the other hand, hard-coding an export in JSON is not going to work well either, as there's no automated way of outputting JSON from the entities -- and doing it by hand is error-prone and hard to maintain.

All in all, outputting the raw pages helps a lot when you make model changes without touching the layout -- you'll easily spot a missing/extra member or a missing class, which otherwise would have gone unnoticed, but unfortunately that's the only scenario where it's really useful.

**Comment by heathermiller** [comment]
That wasn't the question I asked. I asked about the _workflow_. You're proposing ambiguous changes to the workflow-- you want to "sample raw files" and hope that people will manually email you diffs? Sounds like it wouldn't really work in practice if arbitrary people (in LAMP/Typesafe or the community) are supposed to know that they're supposed to be personally emailing you diffs from time to time on some arbitrary sampling of files. And it doesn't sound very thorough if you expect sampling a few files to find errors for you. (Think about the errors you found  which lead to you wanting to add `DocTemplateImpl`, `NoDocTemplateMemberImpl` and `NoDocTemplateImpl`. There were only a handful of places where some synthetically-generated entity didn't have a page that should've had one, right? I don't think that sampling a small number of files would have ever enabled you to find these.)

It's also just wrong, in the interest of a productive and healthy workflow, to have to update a bunch of failing tests each time you make a trivial change to anything in scaladoc.

That said, I'm still fully against scaladoc tests, and I don't think this solution really solves much (it helps in probably 10% of cases, max). It's also not what we all discussed and agreed to a while back. I still think that the way tests are handled doesn't make sense at all-- this is evident in the somewhat accepted behavior of just disabling or commenting-out failing scaladoc tests when they pop up. 

You also didn't explain how you believe JSON is error prone. I don't agree at all (you stated it twice, but didn't validate why you think it's more error-prone than plain text). If you're saying that the language is changing too fast, and that we might add a keyword or something that would require someone to have to change the JSON parser in scaladoc, then that's not a very strong argument. This is actually a scenario where it's not totally blasphemous to manually update three dozen failing tests by hand (rather than how we currently do it-- manually update three dozen failing tests by hand each time you touch scaladoc).

> All in all, outputting the raw pages helps a lot when you make model changes without touching the layout -- you'll easily spot a missing/extra member or a missing class, which otherwise would have gone unnoticed, but unfortunately that's the only scenario where it's really useful.

How realistic is this to have to run every night and on every check-in build? It sounds worthwhile only when you, Vlad, are fixing the model (as I'm unaware of anyone else these days who touches the model). It sounds like a really bad idea to run these so often, though, as you're essentially forcing everyone who might touch scaladoc to have to deal with tests that are all-too-sensitive about anything else except for the model that only you touch. That doesn't make sense. This makes sense for _manual_ testing (you on your own), but not nightly and frequent _automatic_ tests.

You're shooting yourself in the foot if you want scaladoc contributors from the community. Why? 95% of people contributing to scaladoc will end up having to touch the front-end, where you'll force people to have to deal with a billion inflexible tests failing.

**Comment by VladUreche** [comment]
All I'm saying is I want a **separate** jenkins job that periodically generates the raw output, does a diff for a couple of important pages (like List) and emails it to me. It won't fail any tests on anyone, it will just enable me to check that the model changes didn't break anything. One example: when Simon added backticks everywhere, he completely broke the usecase types and linking -- if I could have seen the extent of the damage I could have reacted quicker to either adapt the model or revert his commit. But if I can't get a bird's eye view of what changed you can't expect me to properly fix it, no? And for you not to worry about it: If layout changes between two job runs, I can just ignore the diff, as long as no model changes took place.

Regarding JSON -- how would you do it? Would you use an automated tool based on reflection, or would you code it up yourself? If (1), you'd be exposing internal fields of the _Impl classes, that's not what you want to test. If (2), you'd need to keep the export code in sync with the model *_by hand**, and that's error prone -- who can check you exported all the fields correctly?

Regarding commenting out scaladoc tests:
- the html tests will be gone at some point, I just don't want to spend two days on rewriting them now (If you have such strong feelings about them, why not spend some time translating a couple of them to model tests, I did that whenever I had failing html tests)
- commenting out a test, either in scaladoc or anywhere else **should not be done.** That's not to say it hasn't been done before, because people just don't want to spend the time figuring out what's failing, but that's not something we should accept, especially since we have the pull request mechanism.

**Comment by VladUreche** [comment]
A note to scala gatekeepers (@adriaanm @jsuereth and @paulp, I guess) -- this pull request got it's confirmation from Heather and we should really include it in M5. So please MergeThxBye this patch before M5! 
KThxBye!

**Pull request merged**